### PR TITLE
make sleep work for much shorter time increments

### PR DESCRIPTION
**Pull request by oscardssmith**
I'm not convinced that this is the best implementation of this, but it is an easy one that will reliably work cross platform.  We should probably introduce higher precision `Timer`s since in the past few years they have gained support in Windows which was the last holdout, but IMO this PR on it's own is worthwhile.

```
julia> @benchmark Base.mysleep(1e-4)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):   99.121 μs … 120.211 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     100.730 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   100.571 μs ± 616.414 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                                          ▁▃ ▃▆▂▅█▂▅▄ ▂▂ ▂       
  ▂▁▁▂▂▂▂▃▃▃▄▅▄▅▅▆▆▇▆█▇▅▆▆▅▅▇▄▄▄▄▄▃▄▄▄▅▆█▆███████████▇██▆██▅▆▅▃ ▅
  99.1 μs          Histogram: frequency by time          101 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark Base.mysleep(1e-5)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  10.120 μs …  15.920 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     10.860 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   10.867 μs ± 284.132 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                     ▂▁▂▃▆▁  ▂      ▁▅        ▃▆▃█▁             
  ▂▁▂▂▁▂▂▂▁▂▂▂▂▂▃▃▄▄███████▇██▅▆▃▂▄▅██▆▄▄▄▆▄▇▆██████▇▇▇▄▄▂▂▂▂▂ ▄
  10.1 μs         Histogram: frequency by time         11.4 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark Base.mysleep(2e-6)
BenchmarkTools.Trial: 10000 samples with 9 evaluations.
 Range (min … max):  2.357 μs …   4.843 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.534 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.541 μs ± 107.653 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

       ▁▃ ▁▂▁▄▄▂       ▃▁▇▂▄▂ ▃         ▂▄█▁▂                  
  ▁▁▂▄███████████▆▄▃▃▅▇██████████▆▃▃▃▄▇▇█████▇█▇▅▄▃▂▁▁▁▁▁▁▁▁▁ ▄
  2.36 μs         Histogram: frequency by time        2.78 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark Base.mysleep(1e-6)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.528 μs …   5.098 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.723 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.742 μs ± 106.515 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                 ▃▇▇██▂▁▅▁▁    ▂▅▇▇▄                           
  ▁▂▂▂▂▂▂▂▃▃▂▂▂▂▄██████████▆▅▄▅█████▆▅▆▆▃▃▃▂▂▄▅▄▅▃▂▂▂▁▁▁▁▁▁▁▁ ▄
  1.53 μs         Histogram: frequency by time           2 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

### CODE REVIEW

**Comment by Seelengrab** [comment]
> We should probably introduce higher precision Timers since in the past few years they have gained support in Windows which was the last holdout

Got a source for that? I'd love to incorporate that API into https://github.com/JuliaLang/julia/pull/45023

File: base/asyncevent.jl [hunk 1]
```diff
@@ -264,13 +264,26 @@ end
...
-Block the current task for a specified number of seconds. The minimum sleep time is 1
-millisecond or input of `0.001`.
+Block the current task for at least specified number of seconds.
+The resolution of sleep is around 2 microsecond (2e-6 seconds), but
+as the amount of time decreases, the likelihood of the sleep exceeding the
+given amount of time increases (since the OS or julia scheduler might not
+wake up the thread/task at exactly the right time).
+
+Note that using sleep for small increments of time usually implies that you are
+doing something weird and should consider using some form of notification system.
```
Comment (Seelengrab) [inline base/asyncevent.jl:274]: "something weird" sounds like an odd judgement call. There are very valid reasons to suspend execution for very short amounts of time, even without any expectation that other tasks run (which `sleep` also doesn't guarantee). Not all such usecases would benefit from a notification system either.

Comment (oscardssmith) [reply base/asyncevent.jl:274]: the reason it's weird is that if you are specifying a very small number, it implies that you want to regain control after a very small amount of time. However, due to the nature of schedulers that is not actually a thing that sleep can guarentee, so you either should be using `yield` if you don't care about not coming back immediately, writing a busy loop if you do, or if you are using it in a loop to check a condition, you should be using an event system of some sort that notifies you when the thing you are waiting for completed (e.g. using FileWatching.jl if you want to get notified when a file is modified).

**Comment by oscardssmith** [comment]
https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getsystemtimepreciseasfiletime looks like <1us timers have existed for at least 10 years.

File: base/asyncevent.jl (hunk 1 above)
Comment (Seelengrab) [reply base/asyncevent.jl:274]: Yes, I know all of that, that's why I said that there are valid usecases for such short sleeps. I'm saying that the wording is judgmental, and that's not appropriate for a docstring. All I'm asking for is to reword this a bit.

**Comment by Seelengrab** [comment]
I haven't measured so I don't know how this actually compares, but that seems to come at a [performance cost](https://devblogs.microsoft.com/oldnewthing/20170921-00/?p=97057). This is OT for this PR, but I'll keep it in mind for the other PR, thank you!

**Review by Seelengrab — APPROVED** [review]

**Comment by vtjnash** [comment]
Windows normally rounds timers up to the nearest 16ms (60hz) interval, to save power. Until recently, it would always round up to the nearest 2ms deadline at minimum. There appears to now be a flag we could pass to disable that rounding behavior, though it may reduce the program throughput performance.

**Comment by oscardssmith** [comment]
@vtjnash I think you're confusing the normal GetSystemTime with GetSystemTimePrecise. The Precise variants use the performance counters to get ~100ns accuracy (at least between timings)

**Comment by gbaraldi** [comment]
https://learn.microsoft.com/en-us/windows/win32/sysinfo/acquiring-high-resolution-time-stamps Microsoft has some docs on getting precise timers

**Comment by LilithHafner** [comment]
From triage: 

Do we really need this? Why do we need this?

The real blocker: this should be at the libuv level, not a hack on `sleep`. At the very least it should support `Timer`s too.

Anyone doing short sleeps should, as the docstring says, probably do something else.

**Comment by Seelengrab** [comment]
> Anyone doing short sleeps should, as the docstring says, probably do something else.

As I mentioned in a review comment above, doing extremely short busy sleeps is appropriate in some situations. Not all of those situations benefit from doing "something else" either, especially if those solutions have either more overhead (`yield`) or aren't architecturally feasible. For example, in [StyledApple.jl](https://github.com/Seelengrab/StyledApple.jl/blob/15e3240468bfdc1169afc6ae71191d0599384102/apple.jl#L57-L60) I really do need a busy loop for alignment to 16ms blocks, since yielding is potentially too expensive if the currently processed frame takes 15.8ms, for example.

**Comment by MilesCranmer** [comment]
I think this shouldn't yet be closed. Also relevant is the discussion https://github.com/JuliaLang/julia/issues/54971 and https://github.com/JuliaLang/julia/issues/12770. IMO short-duration sleep is a fundamental feature of a programming language. High level languages like Python can do this. Low level languages like C can do this. It is available on many operating systems at a low level and I think Julia should make it accessible to the user.

Having users "do something else," in effect, means refactoring codebases around the task API – this restructuring can be awkward, overkill, and bug-inducing for many use-cases compared to a simple short-duration sleep in a loop.

**Comment by JeffBezanson** [comment]
To clarify the summary from triage: we definitely believe this should be implemented, but it should be done by improving libuv and timers, not just hacking it into `sleep` (which is just a simple wrapper around timers) doing busy waiting.

**Comment by JeffBezanson** [comment]
> since yielding (or even some form of notification system, which amounts to much the same thing) is potentially too expensive

If that's the case I think writing your own busy loop is the correct solution; `sleep` obviously will yield in general and I'm not sure we should guarantee that it busy loops at a particular threshold.

> Of course, the "ideal" solution for that problem is running my own scheduler, but that's not at all feasible with the currently exposed Task-based API.

I'm just curious about this. Why isn't it feasible? You can manually select tasks to run (`yieldto` or `yield(t)`) and switch to them without going through our scheduler, or pin tasks to threads.

**Comment by Seelengrab** [comment]
> sleep obviously will yield in general

If that's the expected behavior, i.e. `sleep` is guaranteed to yield to the scheduler, that should be documented. As is, the docstring doesn't specify one way or another, and other languages do the exact same thing as this PR would 🤷

> Why isn't it feasible? You can manually select tasks to run (yieldto or yield(t)) and switch to them without going through our scheduler, or pin tasks to threads.

I can't prevent tasks from yielding to the julia scheduler when they're yielding themselves, e.g. when they encounter a lock or perform I/O and need to wait on libuv, or simply call `sleep` themselves. I could provide custom implementations of all of that, but that still doesn't prevent unaware library code from yielding to the wrong scheduler when they're yielding themselves.

This is also one of the reasons why I haven't added testing of threaded statemachines to Supposition.jl, since the julia scheduler is not deterministic in a controllable way. See [this article](https://seelengrab.github.io/articles/Antithesis%20is%20not%20enough/) on my blog for why I'd love to have that.

**Comment by StefanKarpinski** [comment]
> If that's the expected behavior, i.e. sleep is guaranteed/required to yield to the scheduler, that should be documented. As is, the docstring doesn't specify one way or another, and other languages do the exact same thing as this PR would 🤷 If we want this behavior but just want to put it into Timer instead of sleep, that's good feedback for the PR author, but IMO not a good reason to outright close a PR.

It's not required to yield but, like every blocking operation in Julia, it may yield if there's other work to do and since we don't do preemption, there's no guarantee of when a task will resume once it yields. This isn't documented for sleep specifically, but it's how every single blocking operation in the language works, which seems pretty clear from the general I/O and scheduling documentation. But perhaps it could be clarified somewhere. The alternative would be that `sleep` is a true blocking call that blocks the thread entirely, which is the wrong thing in 99% of situations. But, like I said, we don't have any of those anywhere and I don't really think we should.

**Comment by Seelengrab** [comment]
> The alternative would be that sleep is a true blocking call that blocks the thread entirely, which is the wrong thing in 99% of situations.

This PR though would only affect the cases where the scheduler overhead/libuv doesn't give better precision anyway, in which case it clearly is the right behavior. The task is blocked for a duration so short that it's not generally worth it to actually switch execution contexts.

> and I don't really think we should.

I have yet to see a good argument for why that is, other than the unspecific "would be better served by other approaches". I've already linked a project of mine where those other approaches aren't a good fit either and upthread people also agree that this is generally a (sometimes needed) good tool.

So I ask again: please consider usecases other than the one you're used to.

**Pull request closed without merging**